src/odometry_publisher.cpp

```diff
@@ -0,0 +1,215 @@
+#include "odometry_publisher.h"
+
+/**
+ * Constructor, initialises variables and node stuff;
+ */
+OdometryPublisher::OdometryPublisher(DPR4Base *baseP, ros::NodeHandle n) {
+    base = baseP;
+
+    x = 0;
+    y = 0;
+    th = 0;
+    leftPos = base->getLeftPos();
+    rightPos = base->getRightPos();
+    leftDistance = 0;
+    rightDistance = 0;
+
+    overflowPoint = 32.768d;
+    overflowThreshold = 50.0d;
+    logging = true;
+
+    lastTime = ros::Time::now();
+    odom_pub = n.advertise<nav_msgs::Odometry>("odom", 50);
+}
+
+/**
+ * Publish the odometry/transform messages
+ */
+void OdometryPublisher::publishOdometry() {
+    this->updateOdometry();
+
+    // Calculate time interval
+    ros::Time curTime = ros::Time::now();
+    double dt = (curTime - lastTime).toSec();
+    lastTime = curTime;
+
+    // No rotation
+    geometry_msgs::Quaternion cam_quat = tf::createQuaternionMsgFromYaw(0.0);
+
+    // Transform for kinect, x is forwards, y left, z upwards.
+    geometry_msgs::TransformStamped camera_trans;
+    camera_trans.header.stamp = curTime;
+    camera_trans.header.frame_id = "base_link";
+    camera_trans.child_frame_id = "camera_depth_frame";
+    camera_trans.transform.translation.x = 0.08;
+    camera_trans.transform.translation.y = 0.0;
+    camera_trans.transform.translation.z = 0.2;
+    camera_trans.transform.rotation = cam_quat;
+
+    odom_broadcaster.sendTransform(camera_trans);
+
+    // Again, no rotation
+    geometry_msgs::Quaternion odom_quat = tf::createQuaternionMsgFromYaw(th);
+
+    // Transform for odom/base_link, x and y inverted because different axis
+    // used in our calculations of odom.
+    geometry_msgs::TransformStamped odom_trans;
+    odom_trans.header.stamp = curTime;
+    odom_trans.header.frame_id = "odom";
+    odom_trans.child_frame_id = "base_link";
+    odom_trans.transform.translation.x = y/100.0;
```

Comment:
Is that really necessary? I think most people remember a meter being 100 centimeters, will comment it though